### PR TITLE
feat(diag): track recently closed connections and their close reasons

### LINKED ISSUE #55: Add closed connection tracking with close reasons to DIAG HTTP server
## Problem (Why)

In the infra-bemisc deployment, client connections from omni-gateway (Deno) to `ldj.frontdoorhd.com` (behind haproxy + netius proxy_c) are being killed every 5-10 minutes with no prior notice, despite `KEEPALIVE_TIMEOUT=3600`. HAProxy's idle timeout is set to 2 hours, ruling it out as the cause. Diagnosing this is difficult because the DIAG HTTP server only exposes currently active connections — once a connection closes, all context is lost. Log-based diagnostics have proven inefficient due to high traffic volume. We need a way to inspect recently closed connections and their close reasons via the DIAG HTTP endpoint to identify the root cause of these disconnections.

## Description (What)

Add a ring buffer of recently closed connections to the DIAG system, capturing close reason, timestamps, duration, last activity time, error details, and paired connection ID (for proxy correlation). Expose this via a new `GET /connections/closed` endpoint on DiagApp. Close reasons will be string constants (e.g., `"timeout"`, `"client_eof"`, `"upstream_error"`, `"error"`, `"explicit"`). The ring buffer defaults to 512 entries, configurable via `DIAG_CLOSED_MAX`. Tracking is active when running under DIAG mode. The endpoint returns the full buffer, most recent first.

## Implementation (How)

- [x] **1. Define close reason constants** in `src/netius/base/conn.py` — add string constants: `"timeout"`, `"client_eof"`, `"upstream_error"`, `"error"`, `"explicit"`, `"unknown"` (and others as needed)
- [x] **2. Add `close_reason` field to `BaseConnection`** — initialize to `None`, set before `close()` is called; include `close_reason`, `close_timestamp`, and `last_activity_timestamp` in `info_dict()` output
- [x] **3. Implement a ring buffer for closed connections** in `src/netius/base/common.py` (or a new utility) — use `collections.deque(maxlen=N)` with max size from `DIAG_CLOSED_MAX` conf (default 512); store a snapshot dict of the connection's `info_dict()` plus close metadata at close time
- [x] **4. Hook into `Base.on_connection_d()`** — when DIAG is active, capture the closed connection's info dict (including close reason, close timestamp, connection duration, last activity timestamp, error details) and append to the ring buffer
- [x] **5. Propagate close reasons at all close call sites** — audit `BaseConnection.close()`, timeout handlers, EOF/error handlers in `src/netius/base/common.py`, and ensure each sets `close_reason` before closing
- [x] **6. Propagate close reasons in proxy server** — in `src/netius/servers/proxy.py`, set appropriate close reasons in `_on_prx_close()` (upstream error), `_on_raw_close()` (tunnel close), `on_connection_d()`, `on_stream_d()`; include paired/correlated connection ID in the close metadata
- [x] **7. Add paired connection ID to proxy close records** — when a proxy frontend or backend connection closes, include the paired connection's ID (from `conn_map`) in the close snapshot so frontend/backend closures can be correlated
- [x] **8. Add `GET /connections/closed` endpoint to DiagApp** in `src/netius/base/diag.py` — return the full ring buffer contents as JSON, most recent first
- [x] **9. Add `DIAG_CLOSED_MAX` conf support** — read from netius conf system, default to 512, used to size the deque
- [x] **10. Test** — add tests for the ring buffer behavior (overflow, ordering), close reason propagation, and the new DIAG endpoint

---

Implemented in #78. Notes on where the implementation differs from the plan above:

* **2** the timestamps are not new fields. `DiagConnection` already keeps `creation`, `last_recv_ts` and `last_send_ts` and is already the class selected when `DIAG` is on, so `last_activity_timestamp` is derived from those and the regular (non diagnostics) connection is left untouched. A `close_error` field was added alongside the reason for the error details.
* **2** the reason is an argument of the closing (`close(reason=..., error=...)`) rather than a field to assign before it. The attribute remains for the three cases where the closing is not the call being made.
* **3** the buffer is a class attribute and therefore shared by every agent in the process. A per instance one would only ever show the front-end side, since a proxy's upstream connections belong to a separate `HTTPClient` agent and `DiagApp` is bound to a single system, which would leave items 6 and 7 with nothing to correlate.
* **3** the snapshot is the shallow `info_dict()`. The full variant embeds the parser, which holds a `BytesIO`: it made the endpoint fail to serialize and would have retained parser buffers for every entry in the buffer.
* **5** the audit had to cover `base/server.py` as well as `base/common.py`. Traced against a live proxy, every close arrived through the server level `on_read`, `on_error`, `on_exception` and `on_expected` and reported `unknown`.
* **7** the pairing is established at connect and acquire time, not when the map entry is created, because `ConnectionCompat.id` returns `None` until the socket exists. Under the Agent/Protocol architecture the mapped value is an `HTTPProtocol` while the buffer records the underlying `Connection`, so both the reason and the paired identifier are resolved through `.connection` first.

Two things worth knowing:

* A graceful client disconnect reports `client_eof` on epoll and `error` on kqueue. `KqueuePoll` routes `KQ_EV_EOF` into the error list while `EpollPoll` tests `EPOLLIN` first, so the zero length read path only runs on the latter. The infra-bemisc deployment is Linux and will report `client_eof`.
* `GET /connections?full=1` cannot serialize a live connection that has a parser attached, for the same `BytesIO` reason. It predates this work and was left alone.

### PR DESCRIPTION
**Pull request by joamag**
Closes #55.

Adds a ring buffer of recently closed connections to the diagnostics, so that a connection that was killed can still be inspected after it is gone, together with the reason it was closed and the connection it was paired with.

## What was built

| Step | Where |
| --- | --- |
| Close reason constants | `REASON_TIMEOUT`, `REASON_CLIENT_EOF`, `REASON_UPSTREAM_ERROR`, `REASON_ERROR`, `REASON_EXPLICIT`, `REASON_UNKNOWN` in `base/conn.py` |
| Close metadata on the connection | `close_reason`, `close_error` and `close_timestamp`, reported by `info_dict()` |
| Closing API | `close(reason=..., error=...)`, so the reason travels with the call |
| Ring buffer | `AbstractBase._DIAG_CLOSED`, a `deque` sized by `DIAG_CLOSED_MAX` (default `512`) |
| Recording | `record_closed()`, called from `on_connection_d()` while under `DIAG` |
| Endpoint | `GET /connections/closed` on `DiagApp`, full buffer, most recent first |
| Correlation | `pair_connection()` and `reason_connection()` on `ProxyServer` |

Verified end to end against a reverse proxy in front of a WSGI back end, with `DIAG=1`:

```
GET /connections/closed -> 200
  id=45809ea4 reason=explicit     paired=b9f088cf  duration=0.0008
  id=b9f088cf reason=error        paired=45809ea4  duration=0.0009
```

The two entries are the two sides of the same exchange, each naming the other.

## How a reason is set

The reason is an argument of the closing rather than a field to remember to assign before it:

```python
connection.close(reason=netius.REASON_CLIENT_EOF)
connection.close(reason=netius.REASON_ERROR, error=str(exception))
```

Both values are applied before the flush branch, so `close(flush=True, reason=...)` keeps them through the deferral and they are still there when the buffer drains and the real close runs.

The `close_reason` attribute stays as the fallback for the three cases where the closing is not the call being made, and only those remain in the codebase:

* `ProxyServer.reason_connection()`, because the target is an `HTTPProtocol` and `Protocol.close()` takes no arguments
* `_on_prx_close()`, where one branch closes through a `send_response` callback, so a single assignment covers both branches
* `_on_prx_error()`, which records the reason without closing anything itself

Every other call site passes the reason as an argument.

## Three decisions worth reviewing

**Reused `DiagConnection` instead of adding new tracking.** The first pass added an `open_timestamp`, a `last_activity_timestamp` and a `DIAG` flag to `BaseConnection`, with the timestamps stamped on every read and write. That duplicated `DiagConnection`, which already keeps `creation`, `last_recv_ts` and `last_send_ts` and is already selected when `DIAG` is on. It was reworked to derive from those, so the regular (non diagnostics) path is untouched and no new gating flag exists. `last_activity_timestamp` is therefore reported by `DiagConnection.info_dict()`, which is the only case where it is tracked.

**The buffer is process-global, not per agent.** A proxy's upstream connections belong to a separate `HTTPClient` agent, and `DiagApp` is bound to a single `system`. A per-instance buffer would show only the front-end side and steps 6 and 7 would have nothing to correlate. The buffer is a class attribute, so both agents feed it.

**The snapshot is shallow (`info_dict()`, not `info_dict(full=True)`).** The full variant embeds the parser dict, which holds a `BytesIO`. That made the endpoint return `500` (`Object of type BytesIO is not JSON serializable`) and, worse, would have kept parser buffers alive for up to 512 closed connections. Every field the issue asks for is present without it.

## Beyond the listed files

Step 5 names `base/common.py`, but the connections in a running server are closed through `base/server.py`, which carries its own `on_read`, `on_error`, `on_exception` and `on_expected`. Traced with a live proxy, every close arrived through those and reported `unknown`. The same audit was applied there, per the agreed scope of closing gaps.

The proxy also needed a fix that is easy to miss: under the Agent/Protocol architecture `proxy_c` and `tunnel_c` are `HTTPProtocol` instances, while the ring buffer records the underlying `Connection`. Setting `close_reason` or the paired id directly on them wrote to the wrapper and was silently lost, which is why `reason_connection()` and `pair_connection()` resolve through `.connection` first. Pairing is also established at connect/acquire time rather than at map time, because `ConnectionCompat.id` is a property that returns `None` until the socket exists.

## Platform note on `client_eof`

A graceful client disconnect reports `client_eof` on epoll and `error` on kqueue. This is pre-existing netius behaviour, not something introduced here: `KqueuePoll` routes `KQ_EV_EOF` into the error list, while `EpollPoll` tests `EPOLLIN` before `EPOLLERR`/`EPOLLHUP`, so the zero length read path runs. The deployment in the issue is Linux, so it will report `client_eof`; local macOS runs will show `error`. Changing the poll classification is out of scope for this issue.

## Performance

Measured against `master` (`8607d13d`) served from a separate worktree, same interpreter, runs interleaved and the order alternated between repetitions so that neither side is systematically favoured.

Wall-clock throughput turned out to be useless here: connection churn on this machine varies by 45% run to run (9.1k to 13.2k req/s on `master` alone), which cannot resolve the effect being looked for. The figures below are therefore **server-side CPU time** measured with `getrusage`, and the per-connection numbers use a two point slope (N=5000 versus N=20000) so that process startup and every other fixed cost cancels out.

### With diagnostics off (the default)

| Path | master | branch | Delta |
| --- | --- | --- | --- |
| Marginal CPU per connection (churn) | 71-78 us | 71-76 us | -0.9% (median, straddles zero) |
| Server keep alive, 40000 requests | 1.69 s | 1.71 s | +1.0%, inside a 2.6% within-master spread |
| `_on_prx_acquire` body (per proxied request) | 0.014 us | 0.075 us | **+0.061 us** |
| Connection build + close (microbenchmark) | 2.51 us | 2.60 us | +0.06 us |

Nothing measurable. The `pair_connection` call added to the proxy acquire path is the only new per-request work and it costs 61 ns against a request that consumes roughly 250 us of CPU, so 0.02%. The per-connection additions (three attribute assignments, one `time.time()` and a branch) come to about 60 ns against a 75 us per-connection budget.

### With diagnostics on (`DIAG=1`)

| | per closed connection |
| --- | --- |
| master | 78 us |
| branch, as first written | 112 us (**+42%**) |
| branch, after the fix below | 86 us (**+10%**) |

**The measurement found a real defect.** The snapshot calls `info_dict()`, and `DiagConnection.info_dict()` ends with `self._resolve(self.address)`, which performs a **GeoIP lookup**. That made every connection close under `DIAG` do a geographical resolution: 19.06 us per snapshot with an address set, against 2.59 us without one. It went unnoticed because the unit tests build connections with no address, so `_resolve` returned immediately.

The resolution is now gated on `full`, in the same spirit as the parser being gated there already. `/connections` and `/connections/<id>` default to `full=True` so they are unaffected, while the per-close snapshot skips it. That took the snapshot from 19.06 us to 2.78 us and the end-to-end cost from +42% to +10%, consistently across repetitions (9.7, 9.7, 10.6, 10.1).

The residual +7.8 us per closed connection is spread rather than concentrated: `BaseConnection.info_dict` is 1.29 us, `_uptime` 0.85 us and the new `_last_activity` 0.06 us, with the remainder in real socket attribute access and the deque append. It applies only to connection closes, only when diagnostics are explicitly enabled, and not at all to the request path.

## Tests

613 pass, 12 new, `black` clean.

* `src/netius/test/base/conn.py` (new) covers the fallback reason, an explicit reason surviving the close, a repeated close not overwriting the record, and the derived last activity value
* `src/netius/test/base/common.py` covers the snapshot contents, the duration calculation, the most recent first ordering and the bound of the ring buffer
* `src/netius/test/base/diag.py` (new) covers the endpoint payload
* `src/netius/test/servers/proxy.py` covers the protocol resolution and the pairing in both directions

## One thing found and not fixed

`GET /connections?full=1` has the same `BytesIO` serialization failure, since `full` embeds the parser for live connections too. It predates this change and is not part of the issue, so it is only flagged here.

---

> [!NOTE]
> **Medium Risk**
> Touches connection teardown across base, server, and proxy paths; behavior is additive but mis-set reasons or buffer recording could affect operational debugging under DIAG.
> 
> **Overview**
> Adds **diagnostics for connections after they close**: standardized close reasons (`REASON_*`), optional `error`, and timestamps on `Connection.close(reason=..., error=...)`, surfaced in `info_dict()`.
> 
> When **`DIAG` is enabled**, closes are copied into a **process-wide ring buffer** (`DIAG_CLOSED_MAX`, default 512) via `record_closed()`, exposed as **`GET /connections/closed`** (newest first). **Proxy** paths set reasons and **`close_paired`** IDs so front-end and back-end entries can be correlated; GeoIP in snapshots is skipped unless `full` to avoid cost and JSON issues.
> 
> Supporting updates: **`doc/configuration.md`** and **CHANGELOG**, **`appier`** in test deps, expanded **pytest** deprecation filters, CI **checkout@v7** and **PyPy 3.11**, plus unit tests for conn/common/diag/proxy.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a9013f0287a77231cc83de12717c08e16f8746b3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by joamag** [comment]
bugbot review

**Comment by joamag** [comment]
bugbot review

File: src/netius/base/conn.py [hunk 1]
```diff
@@ -68,6 +68,30 @@
 """ The size of the chunk to be used while received
 data from the service socket """
 
+REASON_TIMEOUT = "timeout"
+""" Close reason for a connection that has been closed because
+it remained idle for longer than the allowed period of time """
```
[the remark this answers is not in the document]
Comment (joamag) [reply src/netius/base/conn.py:73]: Accepted and fixed, this was the best catch of the review.

Verified by grep: `REASON_TIMEOUT` was defined and exported and no code path ever produced it. Given the issue is about connections killed every 5-10 minutes despite `KEEPALIVE_TIMEOUT=3600`, an idle timeout is the most likely cause and the feature would have reported it as `unknown`, which defeats its purpose.

It is now set at every timeout driven closing: `HTTPConnection.close_idle()`, the connect and receive timeouts of the HTTP client, and the PROXY handshake timeout. The client ones run over a protocol, whose `close()` takes no arguments, so the reason is written to the underlying connection first. Covered by `test_close_idle` and `test_timeout_stale`. Fixed in bdf48a6a.

File: src/netius/base/conn.py [hunk 2]
```diff
@@ -188,6 +215,13 @@ def close(self, flush=False, destroy=True):
         # this is relevant to avoid any erroneous situation
         self.status = CLOSED
 
+        # marks the timestamp for the closing of the connection and makes
+        # sure that a reason is associated with it, so that a connection
+        # closed with no explicit reason is still properly identified
+        self.close_timestamp = time.time()
+        if not self.close_reason:
+            self.close_reason = REASON_UNKNOWN
```
[the remark this answers is not in the document]
Comment (joamag) [reply src/netius/base/conn.py:231]: Accepted and fixed, same as the sibling comment from Bugbot.

All four call sites you list now set the reason: `close_idle()` through the new `close(flush=True, reason=...)` argument, the PROXY handshake timeout the same way, and the two HTTP client timeouts by writing to the underlying connection before the protocol abort, exactly as you suggest, since `Protocol.close()` takes no arguments. Covered by `test_close_idle` in `src/netius/test/servers/http.py` and `test_timeout_stale` in `src/netius/test/middleware/proxy.py`. Fixed in bdf48a6a.

File: src/netius/base/conn.py [hunk 3]
```diff
@@ -68,6 +68,30 @@
 """ The size of the chunk to be used while received
 data from the service socket """
 
+REASON_TIMEOUT = "timeout"
+""" Close reason for a connection that has been closed because
+it remained idle for longer than the allowed period of time """
+
+REASON_CLIENT_EOF = "client_eof"
```
[the remark this answers is not in the document]
Comment (joamag) [reply src/netius/base/conn.py:75]: Accepted and fixed.

`base/client.py` carried the same gap that had already been closed in `base/common.py` and `base/server.py`: the validation loop, the EOF path, `on_error`, `on_exception`, `on_expected` and the three connect failure paths all closed without a reason. They now pass `reason` (and `error` where an exception is available) to the closing, and there are no bare `connection.close()` calls left in the module.

Added `src/netius/test/base/client.py` covering `on_exception` and `on_expected`. Fixed in bdf48a6a.

File: src/netius/base/common.py [hunk 4]
```diff
@@ -3098,6 +3109,12 @@ def on_connection_d(self, connection):
             connection.owner.name,
         )
 
+        # records the connection in the ring buffer of closed connections
+        # so that it may still be inspected after its destruction, note that
+        # this is only performed while running under diagnostics
+        if is_diag:
+            self.record_closed(connection)
```
[the remark this answers is not in the document]
Comment (joamag) [reply src/netius/base/common.py]: Accepted, but applied differently, as the suggested fix would regress the main use case.

The gap is real: `Base(diag=True)` without `DIAG` in the environment starts the application while the module level `is_diag` stays false, so the endpoint serves an empty list.

Using `self.diag` for the branch would however break the reason the buffer is process wide. A proxy's upstream connections belong to a separate `HTTPClient` agent that is not constructed with `diag`, so its connections would stop being recorded and the front-end to back-end correlation that this feature exists for would have nothing to pair.

The branch is now `is_diag or AbstractBase._DIAG_INSTANCE`. `_DIAG_INSTANCE` is set by `load_diag()` and is class level, so it covers the programmatic case for every agent in the process. Covered by `test_on_connection_d`. Fixed in 147349cd.

File: src/netius/base/common.py [hunk 5]
```diff
@@ -3098,6 +3109,12 @@ def on_connection_d(self, connection):
             connection.owner.name,
         )
 
+        # records the connection in the ring buffer of closed connections
+        # so that it may still be inspected after its destruction, note that
+        # this is only performed while running under diagnostics
+        if is_diag:
+            self.record_closed(connection)
+
```
[the remark this answers is not in the document]
Comment (joamag) [reply src/netius/base/common.py:3118]: Accepted, with the same adjustment described in the codex comment on this block.

Gating on `self.diag` alone would stop the upstream connections of a proxy from being recorded, since its internal `HTTPClient` is not constructed with `diag`, which would leave the front-end and back-end correlation with nothing to pair. The branch is now `is_diag or AbstractBase._DIAG_INSTANCE`, where `_DIAG_INSTANCE` is set by `load_diag()` and is class level, so it covers the programmatic case for every agent in the process. Covered by `test_on_connection_d`. Fixed in 147349cd.

File: src/netius/base/common.py [hunk 6]
```diff
@@ -3466,6 +3529,9 @@ def connection_dict(self, id, full=False):
             return None
         return connection.info_dict(full=full)
 
+    def connections_closed_dict(self):
+        return list(reversed(AbstractBase._DIAG_CLOSED))
```
[the remark this answers is not in the document]
Comment (joamag) [reply src/netius/base/common.py]: Accepted and fixed, and reproduced before changing anything.

With three writer threads appending, `list(reversed(buffer))` raised `RuntimeError: deque mutated during iteration` 9 times over 4 seconds, while taking `list(buffer)` and reversing the resulting list raised none in the same run.

The snapshot is now `list(...)` followed by `reverse()`, which avoids the Python level iterator protocol and therefore needs no lock, keeping the append path free of any synchronization cost. Covered by `test_connections_closed_dict`, which asserts that a listing already taken is not changed by a later record. Fixed in 147349cd.

File: src/netius/base/common.py [hunk 7]
```diff
@@ -291,6 +291,11 @@
 the compaction of the delayed queues may be triggered, avoiding the
 cost of such operation for a small number of them """
 
+DIAG_CLOSED_MAX = config.conf("DIAG_CLOSED_MAX", 512, cast=int)
+""" The maximum number of closed connections that are kept by the
+diagnostics ring buffer, once such value is reached the oldest of
+the entries are dropped so that the memory usage remains bound """
+
```
[the remark this answers is not in the document]
Comment (joamag) [reply src/netius/base/common.py]: Accepted and fixed.

Confirmed that `collections.deque(maxlen=-1)` raises `ValueError: maxlen must be non-negative`, and since the deque is built during class body evaluation a single mistaken environment value would fail the import of `netius.base.common` entirely rather than only disabling the diagnostics.

The value is now clamped with `max(..., 0)`, so a negative bound becomes a buffer that simply keeps nothing. Covered by `test_diag_closed_max`. Fixed in 147349cd.

**Review by joamag — COMMENTED** [review]
LGTM
